bee-ledger/src/workers/consensus/white_flag.rs

```diff
@@ -68,54 +58,190 @@ fn apply_regular_essence<B: StorageBackend>(
                     return Ok(ConflictReason::InputUtxoNotFound);
                 }
             }
-            Input::Treasury(_) => {
+            _ => {
                 return Err(Error::UnsupportedInputKind(input.kind()));
             }
         };
 
-        let _essence_hash = TransactionEssence::from(essence.clone()).hash();
+        consumed_outputs.push((*output_id, consumed_output));
+    }
 
-        match consumed_output.inner() {
-            Output::Treasury(_) => return Err(Error::UnsupportedOutputKind(consumed_output.inner().kind())),
-            Output::Basic(_) => {
-                // TODO
-            }
-            Output::Alias(_) => {
-                // TODO
-            }
-            Output::Foundry(_) => {
-                // TODO
+    let inputs: Vec<(OutputId, &Output)> = consumed_outputs
+        .iter()
+        .map(|(output_id, created_output)| (*output_id, created_output.inner()))
+        .collect();
+
+    let mut context = ValidationContext::new(
+        transaction_id,
+        essence,
+        inputs.iter().map(|(output_id, input)| (output_id, *input)),
+        unlock_blocks,
+        metadata.milestone_index,
+        metadata.milestone_timestamp,
+    );
+
+    // Validation of the inputs commitment.
+    if context.essence.inputs_commitment() != &context.inputs_commitment {
+        return Ok(ConflictReason::InputsCommitmentsMismatch);
+    }
+
+    // Validation of inputs.
+    for ((output_id, consumed_output), unlock_block) in inputs.iter().zip(unlock_blocks.iter()) {
+        let (conflict, amount, consumed_native_tokens, unlock_conditions) = match consumed_output {
+            Output::Basic(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Alias(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Foundry(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Nft(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
```

Comment:
You could introduce an `OutputEssence` (or differently named) trait and implement it for all output types, so that the compiler knows that `unlock(..)`, `amount()`, etc are available for each of those. Then - I think - you can reduce the noise to:

```rust
let ... = match consumed_output {
    use Output::*;
    Basic(output) | Alias(output) | Foundry(output) | Nft(output) => {
         <output as OutputEssence>.unlock(output_id, unlock_block, &inputs, &mut context),
         ...
    },
   _ => ...
}
```

Reply:
All the bindings in the same arm must have the same type so this kind of match is not legal